src/lib/HasuraGraphiQL.tsx

```diff
@@ -0,0 +1,563 @@
+    if (headersInput !== headers) {
+      setSchema(undefined);
+      setLoading(true);
+      setHeaders(headersInput);
+    }
+  };
+
+  const graphQLFetcher = makeFetcher(
+    url,
+    url.replace("http", "ws"),
+    transformHeaders(headers),
+    async function customFetch(...args) {
+      setIsCached(hasCacheDirective(query));
+      setResponseTime(null);
+      let start = Date.now();
+      let returnedPromise = fetch(...args);
+      let res = await returnedPromise;
+      setResponseTime(Date.now() - start);
+      let cloned = res.clone();
+      setResponseSize(JSON.stringify(await cloned.json()).length * 2);
+      return returnedPromise;
+    },
+    () => {
+      setResponseTime(null);
+      return null;
+    }
+  );
+
+  function ErrorNotification() {
+    return (
+      <div className="hasura-graphiql-notifications-wrapper">
+        <div className="hasura-graphiql-notification-tr">
+          <div className="hasura-graphiql-notification-inner">
+            <h4 className="hasura-graphiql-notification-title">
+              Schema Introspection Error
+            </h4>
+            <div className="hasura-graphiql-notification-message">
+              {error ? error : "Error loading schema"}
+            </div>
+            <span
+              className="hasura-graphiql-notification-dismiss"
+              onClick={() => setErrorShown(false)}
+            >
+              ×
+            </span>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  const extraButtons = () => {
+    const buttons = [
+      {
+        label: "Explorer",
+        title: "Toggle Explorer",
+        onClick: () => setExplorerVisible(!explorerVisible),
+      },
+      {
+        label: "Code Exporter",
+        title: "Toggle Code Exporter",
+        onClick: () => setCodeExporterVisible(!codeExporterVisible),
+      },
+    ];
+    if (isCloud)
+      buttons.push({
+        label: "Cache",
+        title: "Cache the response of this query",
+        onClick: () => setQuery(toggleCacheDirective(query)),
+      });
+    return buttons.map((b) => {
+      return <GraphiQL.Button key={b.label} {...b} />;
+    });
+  };
+
+  React.useEffect(() => {
+    if (loading)
+      fetch(url, {
+        method: "post",
+        headers: transformHeaders(headers),
+        body: JSON.stringify({
+          query: getIntrospectionQuery(),
+        }),
+        credentials: "omit",
+      })
+        .then((res) => res.json())
+        .then((data) => {
+          if (data.errors) setError(data.errors[0].message);
+          else setError(null);
+          setSchema(buildClientSchema(data.data));
+          setLoading(false);
+          setErrorShown(false);
+        })
+        .catch(() => {
+          setErrorShown(true);
+          setSchema(undefined);
+          setLoading(false);
+        });
+  }, [headers, url, loading]);
+
+  return (
+    <div id="hasura-graphiql-wrapper">
+      {errorShown && <ErrorNotification />}
```

Comment:
I think we have different mental models of react hooks. To me, when using hooks, react is an engine that updates the UI when data changes. UI follows data. Only after the data and UI are synchronised, the state variable values make sense. So, it's acceptable to have a state variable like `errorShown` - because after the UI is synchronised, the data says `errorShown:true` and the UI shows the error.

Looks like others too have a similar mental model and name state variables in past tense - Eg:

https://github.com/excalidraw/excalidraw/blob/afa7932c9b25980534012b1ef85a6e9b5743b95c/src/components/ErrorDialog.tsx#L14

https://github.com/getsentry/sentry/blob/6d212ce44cd87b9f9751d4b6beae872da5c953e1/static/app/views/alerts/list/index.tsx#L52

Also, given this mental model, I would not name a state variable as `shouldShowError`. It sounds like an instruction and wouldn't belong in data.

Finally, I think a name like `errorPaneVisible` would make sense irrespective of the mental model. Do let me know if that's acceptable.

Reply:
Yeah I like `errorPaneVisible` too.
How I look at it is: "Present" is when the code is being executed, so the variables should cater to that "present". `errorPaneVisible` and `shouldShowError` fits in that mental model, so feels right to me. But yeah, all of us prolly differ in how we visualise the React life cycle.